### PR TITLE
feat: add exponential backoff to outgoing requests

### PR DESCRIPTION
**Pull request by dmbrooke**
UT for retry mechanism is in the works (still figuring out mocking responses).

### CODE REVIEW

**Review by olamothe — APPROVED** [review]

File: src/push_api_clientpy/platformclient.py [hunk 1]
```diff
@@ -1,10 +1,14 @@
 from .document import Document, SecurityIdentityType
 from dataclasses import asdict, dataclass
-from typing import Literal, TypedDict
+from typing import Literal
 import requests
+from requests.adapters import HTTPAdapter, Retry
```
[the remark this answers is not in the document]
Comment (dmbrooke) [reply src/push_api_clientpy/platformclient.py:5]: No, this is a part of the `requests` library, which is already installed. To confirm, I re-cloned this branch (to full clear the various caches) and re-setup and re-ran the tests and everything worked as expected ✅

File: src/push_api_clientpy/platformclient.py [hunk 2]
```diff
@@ -103,11 +107,26 @@ class BatchUpdateDocuments:
...
         self.apikey = apikey
         self.organizationid = organizationid
+        self.backoff_options = backoff_options
+
+        self.retries = Retry(status=self.backoff_options.max_retries,
+                        status_forcelist=[429],
+                        backoff_factor=self.backoff_options.time_multiple,
+                        backoff_jitter=self.backoff_options.retry_after
+                        )
+        session.mount('http://', HTTPAdapter(max_retries=self.retries))
```
Comment (mrrajamanickam-coveo) [inline src/push_api_clientpy/platformclient.py:128]: Curious, what does this do? 
Should it be `http://` or `https://`?

**Review by mrrajamanickam-coveo — CHANGES_REQUESTED** [review]

File: src/push_api_clientpy/platformclient.py (hunk 2 above)
Comment (dmbrooke) [reply src/push_api_clientpy/platformclient.py:128]: _**Whoops**_, yes it needs to be `https`! Otherwise the retry configuration won't take affect 😅

**Review by mrrajamanickam-coveo — APPROVED** [review]

**Review by mrrajamanickam-coveo — APPROVED** [review]

File: src/push_api_clientpy/platformclient.py [hunk 3]
```diff
@@ -103,11 +107,26 @@ class BatchUpdateDocuments:
...
+@dataclass
+class BackoffOptions:
+    retry_after: int = DEFAULT_RETRY_AFTER
+    max_retries: int = DEFAULT_MAX_RETRIES
+    time_multiple: int = DEFAULT_TIME_MULTIPLE
+
 
 class PlatformClient:
-    def __init__(self, apikey: str, organizationid: str):
+    def __init__(self, apikey: str, organizationid: str, backoff_options: BackoffOptions, session = requests.Session()):
```
Comment (btaillon) [inline src/push_api_clientpy/platformclient.py:118]: Unless I'm reading this wrong, this looks like it's a breaking change?

Should it be this instead (just like `Source`)?
```suggestion
    def __init__(self, apikey: str, organizationid: str, backoff_options: BackoffOptions = BackoffOptions(), session = requests.Session()):
```

**Review by btaillon — CHANGES_REQUESTED** [review]

File: src/push_api_clientpy/platformclient.py [hunk 4]
```diff
@@ -103,11 +107,26 @@ class BatchUpdateDocuments:
...
         self.apikey = apikey
         self.organizationid = organizationid
+        self.backoff_options = backoff_options
+
+        self.retries = Retry(status=self.backoff_options.max_retries,
+                        status_forcelist=[429],
+                        backoff_factor=self.backoff_options.time_multiple,
+                        backoff_jitter=self.backoff_options.retry_after
```
Comment (dmbrooke) [inline src/push_api_clientpy/platformclient.py:126]: It turns out that this documented field doesn't actually exist on the latest version of urllib3 🤦 Investigating work-arounds now...

Comment (dmbrooke) [reply src/push_api_clientpy/platformclient.py]: EDIT: That didn't seem to resolve the problem. Seeing as this is a known limitation within a very widely used library, it's not deal-breaking. I'll remove time_multiple support and move on.

File: src/push_api_clientpy/platformclient.py (hunk 3 above)
Comment (dmbrooke) [reply src/push_api_clientpy/platformclient.py]: Oh yes, good catch! I applied this suggestion ✅

**Review by y-lakhdar — APPROVED** [review]

**Review by btaillon — APPROVED** [review]

**Pull request merged**